### ISSUE TITLE
canal calico-kube-controllers crashloops on AWS RHeL 8

### ISSUE THREAD

**Issue by dmitry-irtegov** [projectcalico/calico#4659]
## Expected Behavior

I am installing Kubernetes cluster on AWS RHeL 8 (RHEL-8.4.0_HVM-20210504-x86_64-2-Hourly2-GP2), using canal as a CNI provider.  I expect it to work normally.

## Current Behavior

calico-kube-controllers pod enters a crashloop with the log message:
```
[root@ip-172-31-64-36 ~]# kubectl logs -n kube-system calico-kube-controllers-78d6f96c7b-qgfdr
2021-06-07 07:47:18.308 [INFO][1] main.go 92: Loaded configuration from environment config=&config.Config{LogLevel:"info", WorkloadEndpointWorkers:1, ProfileWorkers:1, PolicyWorkers:1, NodeWorkers:1, Kubeconfig:"", DatastoreType:"kubernetes"}
W0607 07:47:18.310155       1 client_config.go:615] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2021-06-07 07:47:18.311 [INFO][1] main.go 113: Ensuring Calico datastore is initialized
[root@ip-172-31-64-36 ~]# kubectl logs -n kube-system calico-kube-controllers-78d6f96c7b-qgfdr
2021-06-07 07:47:18.308 [INFO][1] main.go 92: Loaded configuration from environment config=&config.Config{LogLevel:"info", WorkloadEndpointWorkers:1, ProfileWorkers:1, PolicyWorkers:1, NodeWorkers:1, Kubeconfig:"", DatastoreType:"kubernetes"}
W0607 07:47:18.310155       1 client_config.go:615] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2021-06-07 07:47:18.311 [INFO][1] main.go 113: Ensuring Calico datastore is initialized
2021-06-07 07:47:28.311 [ERROR][1] client.go 261: Error getting cluster information config ClusterInformation="default" error=Get "https://10.96.0.1:443/apis/crd.projectcalico.org/v1/clusterinformations/default": context deadline exceeded
2021-06-07 07:47:28.311 [FATAL][1] main.go 118: Failed to initialize Calico datastore error=Get "https://10.96.0.1:443/apis/crd.projectcalico.org/v1/clusterinformations/default": context deadline exceeded
```
coredns pods also enter a crashloop with log messages
```
[root@ip-172-31-64-36 ~]# kubectl logs -n kube-system coredns-558bd4d5db-bhtwd
.:53
[INFO] plugin/reload: Running configuration MD5 = db32ca3650231d74073ff4cf814959a7
CoreDNS-1.8.0
linux/amd64, go1.15.3, 054c9ae
I0607 07:59:05.361392       1 trace.go:205] Trace[2019727887]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156 (07-Jun-2021 07:58:35.360) (total time: 30000ms):
Trace[2019727887]: [30.000389603s] [30.000389603s] END
E0607 07:59:05.361419       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Failed to watch *v1.Namespace: failed to list *v1.Namespace: Get "https://10.96.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.96.0.1:443: i/o timeout
```
Tests with more complex setup, e.q, Ubuntu master and RHeL 8 worker show different (but likely related) behavior:
calico pods start normally, but coredns pods crashloop.  The investigation show that pod IPs of pods on RHeL 8 node are not accessible from the node itself  (but accessible from other nodes).  Thus, kubelet cannot run a healthcheck on a coredns and kills the pod (though internally the pod is healthy).

This is different from https://github.com/projectcalico/calico/issues/2818 , because:
1.  Problem is specific to an AWS RHeL 8, probably even to a specific AMI, and is not reproduced on other RHeL-based systems in other environments, or non-RHeL AWS images (e.g. AWS Ubuntu 20.4).
2. On a given image it is very reproducible, does not resolve itself (we waited for 6 hours) and is not resolved by pod restarts and other simple methods.
3. This happens only with canal.  Pure calico or pure flannel on the same OS do not have this problem.

## Steps to Reproduce (for bugs)

1. Deploy a RHeL 8 node using AWS AMI RHEL-8.4.0_HVM-20210504-x86_64-2-Hourly2-GP2
2. Install docker, kubeadm and other stuff according to https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/
3. Observe a calico-kube-controllers and coredns pods crasloop.

Snippet from the bash_history:

```
cat <<EOF | sudo tee /etc/modules-load.d/k8s.conf
br_netfilter
EOF
cat <<EOF | sudo tee /etc/sysctl.d/k8s.conf
net.bridge.bridge-nf-call-ip6tables = 1
net.bridge.bridge-nf-call-iptables = 1
EOF
sudo sysctl --system
sudo yum-config-manager     --add-repo     https://download.docker.com/linux/centos/docker-ce.repo
sudo yum install docker-ce docker-ce-cli containerd.io
cat <<EOF | sudo tee /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-\$basearch
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
exclude=kubelet kubeadm kubectl
EOF
sudo setenforce 0
sudo sed -i 's/^SELINUX=enforcing$/SELINUX=permissive/' /etc/selinux/config
sudo yum install -y kubelet kubeadm kubectl --disableexcludes=kubernetes
service docker start
systemctl enable docker
systemctl enable kubelet
service kubelet start
kubeadm init --pod-network-cidr=192.168.0.0/16
curl https://docs.projectcalico.org/manifests/canal.yaml -O
export KUBECONFIG=/etc/kubernetes/admin.conf
kubectl apply -f canal.yaml
```

## Your Environment

* Calico version v3.19.1, flannel:v0.13.0
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes 1.21.1 installed by kubeadm
* Operating System and version: Red Hat Enterprise Linux release 8.4 (Ootpa), Linux ip-172-31-64-36.ec2.internal 4.18.0-305.el8.x86_64
* Link to your project (optional):

**Comment by dmitry-irtegov** [comment]
I managed to catch a coredns pod in "Running" state, and yes, it is not pingable from the node:
```
[root@ip-172-31-64-36 ~]# kubectl get pods -n kube-system
NAME                                                   READY   STATUS             RESTARTS   AGE
calico-kube-controllers-78d6f96c7b-qgfdr               0/1     CrashLoopBackOff   14         52m
canal-8ktcl                                            2/2     Running            0          52m
coredns-558bd4d5db-bhtwd                               0/1     Running            14         53m
coredns-558bd4d5db-wd5l4                               0/1     Running            14         53m
etcd-ip-172-31-64-36.ec2.internal                      1/1     Running            0          53m
kube-apiserver-ip-172-31-64-36.ec2.internal            1/1     Running            0          53m
kube-controller-manager-ip-172-31-64-36.ec2.internal   1/1     Running            0          53m
kube-proxy-9qg96                                       1/1     Running            0          53m
kube-scheduler-ip-172-31-64-36.ec2.internal            1/1     Running            0          53m
[root@ip-172-31-64-36 ~]# kubectl get pod -n kube-system coredns-558bd4d5db-bhtwd -o yaml | grep podIP
    cni.projectcalico.org/podIP: 192.168.0.3/32
    cni.projectcalico.org/podIPs: 192.168.0.3/32
  podIP: 192.168.0.3
  podIPs:
[root@ip-172-31-64-36 ~]# ping 192.168.0.3
PING 192.168.0.3 (192.168.0.3) 56(84) bytes of data.
^C
--- 192.168.0.3 ping statistics ---
7 packets transmitted, 0 received, 100% packet loss, time 6149ms
```

**Comment by manuelbuil** [comment]
I think we are hitting the same problem ==> https://github.com/projectcalico/calico/issues/4662

**Comment by dmitry-irtegov** [comment]
> I think we are hitting the same problem ==> #4662

Yes, stopping the nm-cloud-setup and nm-cloud-setup.timer services and deleting the ip rule (in this order) helped.
Simple deleting the rule does not work, it is recreated in matter of minutes.

However, I'm afraid, these services have some purpose, are you sure disabling them will not have bad effects?

**Comment by manuelbuil** [comment]
> > I think we are hitting the same problem ==> #4662
> 
> Yes, stopping the nm-cloud-setup and nm-cloud-setup.timer services and deleting the ip rule (in this order) helped.
> Now, nothing crashloops and everything is pingable.
> 
> Simple deleting the rule does not work, it is recreated in matter of minutes.
> 
> However, I'm afraid, these services have some purpose, are you sure disabling them will not have bad effects?

Unfortunately no, I am not sure. That is why I reported the issue to the NetworkManager guys to understand what would be an acceptable configuration that will not break things

**Comment by dmitry-irtegov** [comment]
As I understand, this is not exaclty a NM issue, it is a conflict between calico and nm-cloud-config.  Conflict needs two sides, may be calico people would look into this?

**Comment by lwr20** [comment]
As it says here: https://docs.projectcalico.org/getting-started/openshift/requirements#node-requirements network-manager and calico can often conflict. 

https://docs.projectcalico.org/maintenance/troubleshoot/troubleshooting#configure-networkmanager has a possible workaround (i.e. configure networkmanager to leave calico interfaces alone).

**Comment by moadqassem** [comment]
@manuelbuil So the up-to-date solution is simply disabling the `nm-cloud-setup` service? I saw that, the NetworkManager fellas have already merged a potential fix. However, the current NetworkManager that RHEL 8.4 is running is `1.32.10` IIRC.

> https://docs.projectcalico.org/maintenance/troubleshoot/troubleshooting#configure-networkmanager

This actually doesn't work, at least on the first boot. The only way I got it working is via cloud-init. disabling the `nm-cloud-setup` and it's timer, in addition to delete the route table in the same script which is really not optimal. 

Never to mention that `ip_tables`module on RHEL 8.4(I guess this also for RHEL 8.0+) isn't enabled by default which means canal will still fail(flannel uses ip_table)

**Comment by supershal** [comment]
@manuelbuil I faced the same issue with RHEL 8.4 in AWS. I see that NetworkManager has a fix in version >1.32.12 . How can I get the updated NetworkManager on RHEL 8.4? The RHEL 8.4 `baseos` repo has only `1:1.32.10-4.el8` version.

**Comment by manuelbuil** [comment]
> @manuelbuil I faced the same issue with RHEL 8.4 in AWS. I see that NetworkManager has a fix in version >1.32.12 . How can I get the updated NetworkManager on RHEL 8.4? The RHEL 8.4 `baseos` repo has only `1:1.32.10-4.el8` version.

I don't know :(. I am disabling it
